evalml/pipelines/binary_classification_pipeline.py

```diff
@@ -60,3 +69,34 @@ def _score(X, y, predictions, objective):
         if predictions.ndim > 1:
             predictions = predictions.iloc[:, 1]
         return ClassificationPipeline._score(X, y, predictions, objective)
+
+    def _compute_predictions(self, X, y, objectives, time_series=False):
+        """Compute predictions/probabilities based on objectives."""
+        y_predicted = None
+        y_predicted_proba = None
+        if any(o.score_needs_proba for o in objectives) or (any(not o.score_needs_proba for o in objectives) and self.threshold is not None):
```

Comment:
Heh, yeah this is confusing stuff, good stuff to clarify. Hopefully this helps (and points out any errors I might have missed). Here's the parred-down version of the code in main for `_predict()`:

```
    def _predict(self, X, objective=None):
        if self.threshold is None:
            return self._component_graph.predict(X)
        ypred_proba = self.predict_proba(X).to_dataframe()
        ypred_proba = ypred_proba.iloc[:, 1]
        if objective is None:
            return infer_feature_types(ypred_proba > self.threshold)
        return infer_feature_types(objective.decision_function(ypred_proba, threshold=self.threshold, X=X))

```

Basically, if we have a `threshold` set for our pipeline, we use `predict_proba` and then either call `ypred_proba > self.threshold` or the decision function if passed an objective.

This logic here for computing the predictions means is trying to say:
1. Calculate predict_proba if any objective needs it for scoring (old behavior)
2. If objective can optimize threshold (aka score doesn't need probability--see `can_optimize_threshold` property) and the threshold has been set, then we will still want the predict probability to calculate the last two lines.

To your confusion @chukarsten:
1. If we only have an objective w/ score_needs_proba: we calculate (conditional is True)
2. If we only have an objective that has an optimizable threshold and threshold has been set: we calculate
3. If we only have an objective that has an optimizable threshold and threshold has not been set: we don't need to calculate (will call `self._component_graph.predict(X)` line above)

Writing this up though, maybe I can simplify this to `if any(o.score_needs_proba for o in objectives) or self.threshold is not None)` since `any(not o.score_needs_proba for o in objectives)` is captured in setting the threshold 🤔

Reply:
@chukarsten 
Okay, trying to wrap my head around this some more, and I guess the benefit of having `if any(o.score_needs_proba for o in objectives) or self.threshold is not None)` is that you only calculate if the threshold is set. In the case where score_needs_proba is False and the threshold is None, you don't need to calculate.


```
score needs proba | threshold is None | calculate predict_proba?
------------------------------------------------------------------------
           T               T                   YES
------------------------------------------------------------------------
           T               F                   YES
------------------------------------------------------------------------
           F               T                   YES
------------------------------------------------------------------------
           F               F                   NO

```

Reply:
@angela97lin Maybe I'm confused now too 😂  but shouldn't the last two 'YES' and 'NO' be flipped? If the threshold is None and the score does not need probability, you don't need to calculate predict_proba?

Overall, I'm in favor of rewritting in terms of `if any(o.score_needs_proba for o in objectives) or self.threshold is not None)`

Reply:
@freddyaboulton I think I'm the confused one, haha--fixed! But sounds good, I'll simplify this 😁